core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -169,70 +284,116 @@ public void segmentsLost(IntSet lostSegments) {
+    * @param concurrentSegments segments map of semgnets left to complete. remove an entry when a segment is completed to
+    *                           prevent a data rehash causing a retry for the given segment
+    * @param <I> input type of the data
+    * @param <R> resulting value
+    * @return Flowable that publishes a result for each segment
+    */
+   protected <I, R> Flowable<R> exactlyOnceParallel(CacheSet<I> set,
+         Set<K> keysToExclude, Function<I, K> toKeyFunction, IntSet segments,
+         Function<? super Publisher<I>, ? extends CompletionStage<R>> transformer,
+         SegmentListener listener, IntSet concurrentSegments) {
+      int extraThread = cpuCount;
+      Flowable<R>[] processors = new Flowable[extraThread + 1];
+      PrimitiveIterator.OfInt segmentIter = segments.iterator();
+      for (int i = 0; i < extraThread; i++) {
+         int initialSegment = getNextSegment(segmentIter);
+         // If the iterator is already exhausted, don't submit to the remaining threads
+         if (initialSegment == -1) {
+            processors[i] = Flowable.empty();
+            continue;
+         }
+         // This is specifically a UnicastProcessor as it allows for queueing of elements before you have subscribed
+         // to the Flowable. It may be worth investigating using a PublishProcessor and eagerly subscribing to avoid
+         // the cost of queueing the results.
+         FlowableProcessor<R> processor = UnicastProcessor.create();
+         processors[i] = processor;
+         asyncScheduler.scheduleDirect(() ->
+               handleParallelSegment(segmentIter, initialSegment, set, keysToExclude, toKeyFunction, transformer, processor,
+                     concurrentSegments, listener));
+      }
+      // After we have submitted all the tasks to other threads attempt to run the segments in our invoking thread
+      int initialSegment = getNextSegment(segmentIter);
+      if (initialSegment != -1) {
+         FlowableProcessor<R> processor = UnicastProcessor.create();
+         processors[extraThread] = processor;
+         handleParallelSegment(segmentIter, initialSegment, set, keysToExclude, toKeyFunction, transformer, processor,
+               concurrentSegments, listener);
       } else {
-         stageFlowable = new FlowableFromIntSetFunction<>(segments, segment -> {
-            Flowable<I> innerFlowable = Flowable.fromPublisher(set.localPublisher(segment))
-                  // If we complete the iteration try to remove the segment - so it can't be suspected
-                  .doOnComplete(() -> concurrentSegments.remove(segment));
+         processors[extraThread] = Flowable.empty();
+      }
 
-            if (keysToExclude != null) {
-               innerFlowable = innerFlowable.filter(i -> !keysToExclude.contains(toKeyFunction.apply(i)));
-            }
+      return ParallelFlowable.fromArray(processors).sequential();
```

Comment:
That is just the rxjava2 operator to combine multiple parallel Flowables down to a single Flowable that only submits element on a single thread at a time. Think you got mixed up about that method on the last PR as well :)

Reply:
I guess what I was asking was why does the result have to submit an element on a single thread at a time, why can't it submit elements in parallel?

Reply:
That is the reactive streams specification Publisher #3 https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.2/README.md#specification

Note that it doesn't state it can't happen on different threads, just not at the same time. You can submit them in parallel you just have to use multiple Publishers then.

Reply:
Aha :)